en/docs/extended_topics/implement_data_loader.md

```diff
@@ -0,0 +1,461 @@
+
+As an example, the data format that Mini Dataloader supported is : A plain text file with two columns of numbers separated by commas (See the `part-000` and `part-001` file in code): 
+
+```text
+1.01,2.02
+2.01,4.02
+3.0,6.05
+4.1,8.205
+5,10
+6.0,12.0
+7.0,14.2
+8.0,16.3
+9.1,18.03
+```
+
+This article will take Mini Dataloader as an example to explain the key points of implementing customized DataLoader. 
+
+# The composition of Dataloader
+
+A complete Dataloader generally includes two types of Op: 
+
+- Data Reader: Which is responsible for loading the data in file system to the input stream of memory and setting the data to the Op's output. 
+- Data Decoder: The Data Decoder decodes and output the data in Data Reader Op. 
+
+For some simple data formats, which is no need for decoding. We can omit the Data Decoder and just use Data Reader. 
+
+As an example, since the data format processed by Mini Dataloader is simple. We still implement the Data Reader and Data Decoder. Among these two Ops: 
+
+- `MiniReader` is responsible for reading data from files and split strings by commas. Convert the text to the float and set to the Op's output. The output shape is two columns of each row. 
+- `MiniDecoder` is responsible for splitting the two columns of each row output in above and get two outputs `x` and `y`, both of them shape is one column of each row. 
+
+In [test_mini_dataloader.py](https://github.com/Oneflow-Inc/oneflow-documentation/tree/master/en/docs/code/extended_topics/data_loader/test_mini_dataloader.py) we can see the use of both Ops at Python level:
+
+ ```python
+    miniRecord = MiniReader(
+        "./",
+        batch_size=batch_size,
+        data_part_num=2,
+        part_name_suffix_length=3,
+        random_shuffle=True,
+        shuffle_after_epoch=True,
+    )
+
+    x, y = MiniDecoder(
+            miniRecord, name="d1"
+        )
+ ```
+
+We will introduce how to implement Data Reader and Data Decoder in C++ backend below. 
+
+# Data Reader operator
+
+## The class relationship in Data Reader
+
+We need to implement a class that inherits from `DataReader`, this class includes two important objects `loader_` and `parser_`, which inherits from `Dataset` and `Parser` separately. 
+
+- `loader_` 's job is to read data from the file system to buffer. The Op's author build the logic by overriding the `Next` function. 
+- `parser_` 's job is to set the data in buffer to Op's output. The Op's author build the logic by overriding the `Parser` function. 
+
+When Data Reader Op works, it will call the relative function in `loader_` to open files in the file system, and then call the `Next` function in `loader_` to read data from the file system according to the logic built by Op's author. 
+
+The pseudocode below shows the class relationship and the calling procedure. The actual code is more complicated than the pseudocode and it's not a one-to-one relationship: 
```

Comment:
The actual code is more complicated and it is not the exact corresponding relationship.